manifest.go

```diff
@@ -831,14 +946,53 @@ func (m *manifestEntryV1) FileSequenceNum() *int64 {
 	return m.FileSeqNum
 }
 
-func (m *manifestEntryV1) DataFile() DataFile { return &m.Data }
+func (m *manifestEntryV1) DataFile() DataFile { return m.Data }
+
+// ManifestEntryV2Builder is a helper for building a V2 manifest entry
+// struct which will conform to the ManifestEntry interface.
+type ManifestEntryV2Builder struct {
+	m *manifestEntryV2
+}
+
+// NewManifestEntryV2Builder is passed all of the required fields and then allows
+// all of the optional fields to be set by calling the corresponding methods
+// before calling [ManifestEntryV2Builder.Build] to construct the object.
+func NewManifestEntryV2Builder(status ManifestEntryStatus, snapshotID int64, data DataFile) *ManifestEntryV2Builder {
+	return &ManifestEntryV2Builder{
+		m: &manifestEntryV2{
+			EntryStatus: status,
+			Snapshot:    &snapshotID,
+			Data:        data,
```

Comment:
since we know that, at least currently, the only implementation of `DataFile` is the internal `dataFile` type, we could simply use a type assert `data.(*dataFile)` to avoid having to use `DataFile` in the manifestlist and builders and instead use the underlying internal type.

Reply:
Happy to make that change, but there are two additional concerns that raises:
1. Not a big deal, but leaving the type as `dataFile` means copying a mutex (`sync.Once`). This shouldn't be an issue in the current usage, but something to consider if you want to avoid that potential issue.
2. Since `NewManifestEntryVXBuilder` accepts an interface type and is exported, a user could potentially provide another type that fullfils the `DataFile` interface. I can sidestep this by changing the signature of the `NewManifestVXBuilder` to return an error on a failed type assertion, but just wanted to bring this up since it muddies the exported footprint. However, accepting and using the interface type means a type that is not avro-encodable will result in a runtime error, so maybe it's a moot point.

Reply:
That's a good point. Though since we're using a `*dataFile` we're avoiding copying right? We're just copying the pointer. 

For the second point, that's a better point which might make it worthwhile to keep things as is then. The primary reason for using an interface was so that we're not exposing the underlying members except via getters which all have to be public in order to allow the avro decoding to work. Given this, I think we should just keep it as is using the interface type directly then.

Reply:
okay - I can revert that change to what I had before. As it stands right now, I switched it to do the type assertion rather than use the interface type inside `manifestEntryVX`. Just to be clear: you'd like to use the interface type in `manifestEntryVX`?

Reply:
okay - all set with that. Can revert the last commit if you'd prefer the other way.